src/crimson/os/seastore/nvmedevice/nvmedevice.h

```diff
@@ -0,0 +1,329 @@
+  crimson::ct_error::input_output_error>;
+
+struct io_context_t {
+  iocb cb;
+  bool done = false;
+};
+
+/*
+ * Interface between NVMe SSD and its user.
+ *
+ * NVMeBlockDevice provides not only the basic APIs for IO, but also helper APIs
+ * to accelerate SSD IO performance and reduce system overhead. By aggresively
+ * utilizing and abstract useful features of latest NVMe SSD, it helps user ease
+ * to get high performance of NVMe SSD and low system overhead.
+ *
+ * Various implementations with different interfaces such as POSIX APIs, Seastar,
+ * and SPDK, are available.
+ */
+class NVMeBlockDevice {
+protected:
+  uint64_t size = 0;
+  uint64_t block_size = 0;
+
+  uint64_t write_granularity = 4096;
+  uint64_t write_alignment = 4096;
+
+public:
+  NVMeBlockDevice() {}
+  virtual ~NVMeBlockDevice() = default;
+
+  template <typename T>
+  static NVMeBlockDevice *create() {
+    return new T();
+  }
+
+  /*
+   * Service NVMe device relative size
+   *
+   * size : total size of device in byte.
+   *
+   * block_size : IO unit size in byte. Caller should follow every IO command
+   * aligned with block size.
+   *
+   * preffered_write_granularity(PWG), preffered_write_alignment(PWA) : IO unit
+   * size for write in byte. Caller should request every write IO sized multiple
+   * times of PWG and aligned starting address by PWA. Available only if NVMe
+   * Device supports NVMe protocol 1.4 or later versions.
+   */
+  uint64_t get_size() const { return size; }
+  uint64_t get_block_size() const { return block_size; }
+  uint64_t get_preffered_write_granularity() const { return write_granularity; }
+  uint64_t get_preffered_write_alignment() const { return write_alignment; }
+
+  virtual read_ertr::future<> read(
+    uint64_t offset,
+    bufferptr &bptr) = 0;
+
+  /*
+   * Multi-stream write
+   *
+   * Give hint to device about classification of data whose life time is similar
+   * with each other. Data with same stream value will be managed together in
+   * SSD for better write performance.
+   */
+  virtual write_ertr::future<> write(
+    uint64_t offset,
+    bufferptr &bptr,
+    uint16_t stream = 0) = 0;
+
+  // TODO
+  virtual int discard(uint64_t offset, uint64_t len) { return 0; }
+
+  virtual open_ertr::future<> open(
+      const std::string& path, 
+      seastar::open_flags mode) = 0;
+  virtual seastar::future<> close() = 0;
+
+  /*
+   * For passsing through nvme IO or Admin command to SSD
+   * Caller can construct and execute its own nvme command
+   */
+  virtual nvme_command_ertr::future<> pass_through_io(
+    NVMePassThroughCommand& command) { return nvme_command_ertr::now(); }
+  virtual nvme_command_ertr::future<> pass_admin(
+    nvme_admin_command_t& command) { return nvme_command_ertr::now(); }
+
+  /*
+   * End-to-End Data Protection
+   *
+   * NVMe device keeps track of data integrity similar with checksum. Client can
+   * offload checksuming to NVMe device to reduce its CPU utilization
```

Comment:
What are the semantics here?  If I write a block using protected_write, is there anything I need to do to validate that checksum, or will any read() (above) work?

Reply:
To read data with SSD internal verification, we should use ioctl(). SSD internal data verification is activated if the PRACT field in the NVMe read command is set to 1---note that the latest kernel nvme driver does not support this feature. So, creating and submitting NVMe IO command manually via ioctl() are the only way to use the End-to-End Data Protection.

Reply:
I mean in terms of this specific interface described in nvmedevice.h.  Is the intention that users use the read() method, or are they supposed to use another to match prototected_write.

Reply:
The latter.

Reply:
This is presumable no longer relevant since this interface will be deferred pending upstream seastar work.

Reply:
Probably some kind of TODO or error indicating that it's not yet been implemented?

Reply:
I think you still need to add a matching protected_read() interface call here, right?

Reply:
> Probably some kind of TODO or error indicating that it's not yet been implemented?
Yeah, after finishing upstream work (ioctl), we will continue TODO.

> I think you still need to add a matching protected_read() interface call here, right?
Right, the interface is added.